### PR TITLE
add ocnn

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?
Adds OCNN for the one class svm task of "novelty detection"
(Please fill in changes proposed in this fix)

## How was this patch tested?
Gradient checks.
(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ ] Reviewed the [Contributing Guidelines](https://github.com/deeplearning4j/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ ] Created tests for any significant new code additions.
- [ ] Relevant tests for your changes are passing.
- [ ] Ran mvn formatter:format (see [formatter instructions](http://code.revelc.net/formatter-maven-plugin/examples.html#Setting_Source_Files) for targeting your specific files).

### CODE REVIEW

File: deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/ocnn/OCNNOutputLayerTest.java [hunk 1]
```diff
@@ -139,6 +143,12 @@ public void testOutput() {
 
         INDArray output = network.output(arr.getRows(ArrayUtil.range(102,149)));
         System.out.println(output);
+        File tmpFile = new File("tmp-ocnn-zip");
+        tmpFile.deleteOnExit();
+        ModelSerializer.writeModel(network,tmpFile,true);
+
+        MultiLayerNetwork loaded = ModelSerializer.restoreMultiLayerNetwork(tmpFile);
+        assertNotNull(loaded);
```
Comment (AlexDBlack) [inline deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/ocnn/OCNNOutputLayerTest.java]: You can replace these 5 lines with a 1-line TestUtils.testModelSerialization(...) call
https://github.com/deeplearning4j/deeplearning4j/blob/b555e1ca4963159e8bb9bcea5c0c162819c7ee0f/deeplearning4j-core/src/test/java/org/deeplearning4j/TestUtils.java#L18

File: deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/ocnn/OCNNOutputLayerTest.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+import org.nd4j.linalg.learning.config.NoOp;
+import org.nd4j.linalg.util.ArrayUtil;
+
+import java.io.File;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class OCNNOutputLayerTest {
```
Comment (AlexDBlack) [inline deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/ocnn/OCNNOutputLayerTest.java]: ```OCNNOutputLayerTest extends BaseDL4JTest```

File: deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/ocnn/OCNNOutputLayerTest.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        int nIn = 4;
+        boolean doLearningFirst = true;
+        MultiLayerConfiguration configuration = new NeuralNetConfiguration.Builder()
+                .seed(42).updater(new NoOp()).miniBatch(false)
+                .list(new  org.deeplearning4j.nn.conf.ocnn.OCNNOutputLayer.Builder().nIn(4)
+                        .nu(0.002)
+                        .nOut(2)
+                        .hiddenLayerSize(numHidden).build())
+                .build();
+        MultiLayerNetwork network = new MultiLayerNetwork(configuration);
```
Comment (AlexDBlack) [inline deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/ocnn/OCNNOutputLayerTest.java]: Add a DenseLayer or something below this... without it: there could be an error in the epsilon (input activation gradients) calculation, and we'd never know it.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/ocnn/OCNNOutputLayer.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+    @Override
+    public ParamInitializer initializer() {
+        return OCNNParamInitializer.getInstance();
+    }
+
+
+    @Override
+    public double getL1ByParam(String paramName) {
+        return l1;
```
Comment (AlexDBlack) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/ocnn/OCNNOutputLayer.java]: r should be excluded from L1 and L2, I would think?

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/ocnn/OCNNOutputLayer.java [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    @Override
+    public INDArray labelProbabilities(INDArray examples) {
+        //LayerWorkspaceMgr.noWorkspacesImmutable()
+        this.input = examples;
+        return doOutput(false,LayerWorkspaceMgr.noWorkspacesImmutable()).subi(getParam(R_KEY));
+    }
+
+    public INDArray getInput() {
```
Comment (AlexDBlack) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/ocnn/OCNNOutputLayer.java]: Unnecessary method given AbstractLayer ```@Data``` annotation?
(And if we do keep it - add override annotation)

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/ocnn/OCNNOutputLayer.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+import org.nd4j.shade.jackson.annotation.JsonCreator;
+import org.nd4j.shade.jackson.annotation.JsonIgnoreProperties;
+import org.nd4j.shade.jackson.annotation.JsonProperty;
+
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * An implementation of one class neural networks from:
+ * https://arxiv.org/pdf/1707.02131.pdf
```
Comment (maxpumperla) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/ocnn/OCNNOutputLayer.java]: Can we use this one? :) I was super confused for a while. I mean, the other paper is interesting as well.

https://arxiv.org/pdf/1802.06360.pdf

File: deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/ocnn/OCNNOutputLayerTest.java (hunk 3 above)
Comment (agibsonccc) [reply deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/ocnn/OCNNOutputLayerTest.java]: See update.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/ocnn/OCNNOutputLayer.java (hunk 4 above)
Comment (agibsonccc) [reply deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/ocnn/OCNNOutputLayer.java]: @AlexDBlack  I don't think so here..I would like the ability to add l2/l1 on top of it. The R is actually part of SGD as a learned parameter.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/ocnn/OCNNOutputLayer.java (hunk 5 above)
Comment (agibsonccc) [reply deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/ocnn/OCNNOutputLayer.java]: fixed

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/ocnn/OCNNOutputLayer.java (hunk 6 above)
Comment (agibsonccc) [reply deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/ocnn/OCNNOutputLayer.java]: done.

**Pull request merged**